### DISCUSSION TITLE
请问大家有没有方法实现脚本的顺序运行

### DISCUSSION THREAD

**Discussion by Phoebe-test** [windingwind/zotero-actions-tags#329] [Q&A]
# 实现脚本的顺序运行
- 功能需求：先运行第一个 Action，再运行第二个 Action
- 问题的提出：部分文章抓取时语言字段为空，因此希望添加新条目时，通过先运行 **Linter** 调整语言字段，再执行**脚本**翻译标题和摘要。两个脚本同时打开可能会有冲突，无法完成功能需求。
## 环境
- OS：Windows 10 22H2 家庭中文版
- Zotero 版本：7.0.0-beta.81+721f54fe4 (64-bit)
- Linter for Zotero 版本：[1.17.0](https://github.com/northword/zotero-format-metadata/releases/tag/v1.17.0)
- Actions and Tags for Zotero 版本：[1.0.0-beta.39](https://github.com/windingwind/zotero-actions-tags/releases/tag/v1.0.0-beta.39)
## 详述：
新建条目后：
&emsp; 1. 通过 Action & Tags 运行 [Linter](https://github.com/northword/zotero-format-metadata/issues/189#issuecomment-2132715266) 调整语言字段
&emsp; 2. 再运行 [zotero-pdf-translate](https://github.com/windingwind/zotero-actions-tags/discussions/107#discussioncomment-7378390) 翻译标题和摘要
以这篇[文章](https://www.nejm.org/doi/10.1056/NEJMoa2001017)为例
| 刚抓取到的界面 | 第一步 | 第二步 |
| --- | --- | --- | 
| ![PixPin_2024-05-27_20-56-49](https://github.com/windingwind/zotero-actions-tags/assets/103734097/36decc0c-6e87-41a3-bf68-58151145e109) | ![PixPin_2024-05-27_20-58-25](https://github.com/windingwind/zotero-actions-tags/assets/103734097/827f476f-3ae0-40a2-a7cf-4d32aacad6a7) | ![PixPin_2024-05-27_21-00-06](https://github.com/windingwind/zotero-actions-tags/assets/103734097/7fa33e0c-530c-4d4e-ba17-b62865aa83e5) |

🙏🙏🙏感谢各位大佬！

**Comment by Phoebe-test** [comment]
# 解决方案：将第二个 Action 写为函数，再延时触发
## 环境
- OS：Windows 10 22H2 家庭中文版
- Zotero 版本：7.0.0-beta.81+721f54fe4 (64-bit)
- Linter for Zotero 版本：[1.17.0](https://github.com/northword/zotero-format-metadata/releases/tag/v1.17.0)
- Actions and Tags for Zotero 版本：[1.0.0-beta.44](https://github.com/windingwind/zotero-actions-tags/releases/tag/v1.0.0-beta.44)

## 说明
- 通过以下两步解决 [脚本与插件的冲突](https://github.com/northword/zotero-format-metadata/issues/189#issuecomment-2132197658)：(Resolve script-plugin conflicts by following these two steps)
1. 先通过 Action & Tags 运行 [Linter](https://github.com/northword/zotero-format-metadata/issues/189#issuecomment-2132715266) 调整语言字段 (First run the plugin Linter through Action & Tags to adjust the language field)
2. 再**延时2000毫秒**运行 [Auto-Translate the New Item's Title](https://github.com/windingwind/zotero-actions-tags/discussions/107#discussioncomment-7378390)翻译标题和摘要 (Then run the script with a delay of 2000 milliseconds to translate the title and summary)

## Action 脚本 (Aciton Script)
- 事件：新建条目 (Event: Create Item)
- 操作：自定义脚本 (Operation: Script)
- 数据： (Data)
```javascript
// 先通过脚本调用插件 Linter (First call the plugin Linter via this script)
// 再延时调用插件 pdf-translate 翻译标题和摘要 (Then delayed call plugin pdf-translate translate title and summary via this script)
// link https://github.com/windingwind/zotero-actions-tags/discussions/329
// Action 中调用 Linter 方法 (Calling Plugin Linter via Action) https://github.com/northword/zotero-format-metadata/issues/189#issuecomment-2132715266

await Zotero.Linter.hooks.onLintInBatch("std", [item]); // 调用插件 Linter (Call the plugin Linter)

const translationResult = await new Promise((resolve) => {
  setTimeout(() => {
    resolve(trans());
  }, 2000);
});
return translationResult; // 弹窗提示翻译结果 (Pop-up window for translation results)

// 将自动翻译标题和摘要的脚本写为函数 (Write scripts for automatic translation of titles and summaries as a function)
async function trans() {
	// Translate item title and abstract automatically
	// Wait for abstract to be filled before translating abstract
	// author windingwind, PhoenixHwang, northword
	// link https://github.com/windingwind/zotero-tag/discussions/107
	// usage Event=Create Item

	const Zotero = require("Zotero");
	if (!Zotero.PDFTranslate) {
		return "[Action:Translate Title and Abstract] Translate for Zotero is not installed or disabled.";
	}
	if (!item) {
		return "[Action:Translate Title and Abstract] Target item is empty";
	}

	const skipLang = ["", "zh", "zh-CN"];
	const lang = item.getField("language");

	if (skipLang.includes(lang)) {
		return "[Action:Translate Title and Abstract] Skip due to language"
	}

	const title = item.getField("title");
	const abstract = item.getField("abstractNote");

	const titleResult = (await Zotero.PDFTranslate.api.translate(title)).result;
	Zotero.PDFTranslate.data.ztoolkit.ExtraField.setExtraField(item, "titleTranslation", titleResult);

	let abstractResult = "";
	if (abstract) {
		abstractResult = (await Zotero.PDFTranslate.api.translate(abstract)).result;
		Zotero.PDFTranslate.data.ztoolkit.ExtraField.setExtraField(item, "abstractTranslation", abstractResult);
	} else {
		const observerID = Zotero.Notifier.registerObserver({
			notify: async (event, type, ids, extraData) => {
				if (event === 'modify' && type === 'item' && ids.includes(item.id)) {
					const newAbstract = item.getField('abstractNote');
					if (newAbstract) {
						abstractResult = (await Zotero.PDFTranslate.api.translate(newAbstract)).result;
						Zotero.PDFTranslate.data.ztoolkit.ExtraField.setExtraField(item, "abstractTranslation", abstractResult);
						Zotero.Notifier.unregisterObserver(observerID);
					}
				}
			}
		}, ['item']);
	}
	
	return `[Action:Translate Title and Abstract] successfully translate title and abstract.`;
}
```

**Reply by desmonna** [reply]
大佬，用这个自动翻译动作如果遇上文献是英文，作者为中式拼音，那么就会跳过翻译，这个怎么修改呢？

**Reply by wzr-ms** [reply]
我也遇到这个识别的问题

**Comment by windingwind** [comment]
@Phoebe-test 

最新版本支持action执行操作为顺序执行多个其他action

**Reply by desmonna** [reply]
请问如何调整动作顺序？

**Reply by windingwind** [reply]
https://github.com/windingwind/zotero-actions-tags?tab=readme-ov-file#action-settings

Operation - otherAction

一行一个依次运行

**Reply by desmonna** [reply]
好的

**Reply by Phoebe-test** [reply]
@windingwind  大佬您好，我更新到了最新版，但似乎并不能很好地顺序执行其他action。已经测试过被触发的两个脚本均能独立运行，并且单独被触发时都没问题，但写在同一个触发脚本中时就有问题，测试文献为：[文献1](https://www.sciencedirect.com/science/article/pii/S0029801823015226?via%3Dihub)、[文献2](https://www.sciencedirect.com/science/article/pii/S002980182402256X)。

还烦请您不吝赐教，环境及脚本详情如下：

## 环境
OS：Windows 10 22H2 家庭中文版
Zotero 版本：7.0.1-beta.3+ec7ffa8c0 (64-bit)
Linter for Zotero 版本：[Release 1.18.1](https://github.com/northword/zotero-format-metadata/releases/tag/v1.18.1)
Actions and Tags for Zotero 版本：[Release v2.0.0](https://github.com/windingwind/zotero-actions-tags/releases/tag/v2.0.0)

## 触发脚本
- 名称：Linter and Translate
- 事件：新建条目
- 操作：触发另一个动作
- 数据：
```
Linter
Translate
```
**测试发现触发脚本的第一行似乎正常执行了，因为语言栏有变化（如图1），但第二行并没有很好的执行（如图2）：**
| 图1 | 图2|
| --- | --- |
| ![无标题](https://github.com/user-attachments/assets/b1c7392e-7e9e-4d1c-aa21-3951a3f33b4b) | ![PixPin_2024-08-19_23-22-44](https://github.com/user-attachments/assets/bbdc9f21-c30a-4484-b0a9-54cc5b280bd8) |


## 被触发脚本1（Linter）
- 名称：Linter
- 事件：无
- 操作：自定义脚本
- 数据：
```
Zotero.Linter.hooks.onLintInBatch("std", [item])
```

## 被触发脚本2（Translate）
- 名称：Translate
- 事件：无
- 操作：自定义脚本
- [数据](https://github.com/windingwind/zotero-actions-tags/discussions/107#discussioncomment-7378390)：
```
// Translate item title and abstract automatically
// Wait for abstract to be filled before translating abstract
// author windingwind, PhoenixHwang, northword
// link https://github.com/windingwind/zotero-tag/discussions/107
// usage Event=Create Item

const Zotero = require("Zotero");
if (!Zotero.PDFTranslate) {
	return "[Action:Translate Title and Abstract] Translate for Zotero is not installed or disabled.";
}
if (!item) {
	return "[Action:Translate Title and Abstract] Target item is empty";
}

const skipLang = ["", "zh", "zh-CN"];
const lang = item.getField("language");

if (skipLang.includes(lang)) {
	return "[Action:Translate Title and Abstract] Skip due to language"
}

const title = item.getField("title");
const abstract = item.getField("abstractNote");

const titleResult = (await Zotero.PDFTranslate.api.translate(title)).result;
Zotero.PDFTranslate.data.ztoolkit.ExtraField.setExtraField(item, "titleTranslation", titleResult);

let abstractResult = "";
if (abstract) {
    abstractResult = (await Zotero.PDFTranslate.api.translate(abstract)).result;
    Zotero.PDFTranslate.data.ztoolkit.ExtraField.setExtraField(item, "abstractTranslation", abstractResult);
} else {
    const observerID = Zotero.Notifier.registerObserver({
        notify: async (event, type, ids, extraData) => {
            if (event === 'modify' && type === 'item' && ids.includes(item.id)) {
                const newAbstract = item.getField('abstractNote');
                if (newAbstract) {
                    abstractResult = (await Zotero.PDFTranslate.api.translate(newAbstract)).result;
                    Zotero.PDFTranslate.data.ztoolkit.ExtraField.setExtraField(item, "abstractTranslation", abstractResult);
                    Zotero.Notifier.unregisterObserver(observerID);
                }
            }
        }
    }, ['item']);
}

return `[Action:Translate Title and Abstract] successfully translate title ${title} to ${titleResult} and abstract.`;
```

**Reply by wzr-ms** [reply]
友友解决这个问题了吗？解决可以分享一下

**Reply by Phoebe-test** [reply]
@wzr-ms  我还是用的之前的脚本QAQ
https://github.com/windingwind/zotero-actions-tags/discussions/329#discussioncomment-9730887

**Reply by wzr-ms** [reply]
谢谢啦，已采用，我的也有这个问题

**Reply by bingofisher** [reply]
可能是每个脚本都是异步函数运行的，所以相当于两个脚本差不多同时运行？
